### ISSUE TITLE
Question about control sequence and thread-safe update without lock

### ISSUE THREAD

**Issue by NoAnyLove** [tqdm/tqdm#430]
- [x] I have visited the [source website], and in particular
  read the [known issues]
- [x] I have searched through the [issue tracker] for duplicates

  [source website]: https://github.com/tqdm/tqdm/
  [known issues]: https://github.com/tqdm/tqdm/#faq-and-known-issues
  [issue tracker]: https://github.com/tqdm/tqdm/issues?q=

The tqdm.update in consists of 3 steps,

1. move to the position
2. print out progress bar
3. move back to the original position

The corresponding code is below,

```
if self.pos:
    self.moveto(self.pos)

# Print bar's update
self.sp(self.__repr__())

if self.pos:
    self.moveto(-self.pos)
```

The main reason that causes the inconsistent output in multi-threading is that the cursor movement and progressbar output maybe interrupted by other threads.

A possible solution without using lock is to combine the cursor positioning control sequence with the progressbar output, and print them out at once.

```python
    @staticmethod
    def status_printer(file, pos):
        """
        Manage the printing and in-place updating of a line of characters.
        Note that if the string is longer than a line, then in-place
        updating may not work (it will print a new line at each refresh).
        """
        fp = file
        fp_flush = getattr(fp, 'flush', lambda: None)  # pragma: no cover

        def fp_write(s):
            fp.write(_unicode(s))
            fp_flush()

        last_len = [0]
        def print_status(s):
            len_s = len(s)
            buf = []
            
            # save the cursor position
            buf.append('\x1b[s')
            # move cursor down
            buf.append('\x1b[%dB' % pos)
            # bar's update
            buf.append('\r' + s + (' ' * max(last_len[0] - len_s, 0)))
            # restore cursor position
            buf.append('\x1b[u')
            # print out at once
            fp_write("".join(buf))
            
            last_len[0] = len_s
        return print_status
```

And we can initialize `self.sp` like this: `self.sp = self.status_printer(self.fp, self.pos)`. Besides, instead of using the save cursor and restore cursor control sequences, we can use Cursor Position sequence `ESC [ <y> ; <x> H` with absolute coordinate within the viewpoint.

By using this approach, we don't need to use `self.moveto` before and after printing the bar, and the tqdm.update will be thread-safe without using any lock. 

What do you think about this approach?

**Comment by casperdcl** [comment]
interesting suggestions, but will it work on as many environments (eg win10 etc) as the current implementation?

**Comment by NoAnyLove** [comment]
@casperdcl I'm not quite sure. I tested on Win8.1 with ConEmu and Ubuntu 16.04, and it works on both environment. Then I realized that the ANSI support for save/restore cursor on my Win8.1 was provided by ConEmu, not by colorama.

I found a reference about ANSI sequence support on Windows, [Console Virtual Terminal Sequences](https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences), I thought the ANSI support was for all windows systems, but finally found that was added in Win10. So the native ANSI sequence support was for Win10, not for eailer windows system.

tqdm currently use colorma to add the support of ANSI sequence on Windows. It now supports cursor positioning `ESC [ y;x H` and `ESC [ y;x f`, the sequences for save/restore cursor positions are not supported yet. But I'm glad to find that there is a pull request to add this feature, tartley/colorama#137.

To summarize, I think it can work on many environments, and we have following choices,

* Use `ESC [ y;x H` and `ESC [ y;x f` with absolute cursor coordinates instead of '\ESC [ s' and '\ESC [ u'. This approach works with current colorma library
* Wait for tartley/colorama#137 to be merged 
* Implement the get and set cursor position functions on windows by using ctypes to invoke Win32 API. The `_environ_cols_windows` function in _utils.py is already a good example. But this breaks our intention of of thread-safe.

**Comment by yelantingfeng** [comment]
I am using tqdm for multiprocess. I specify a different position for each process. When the codes are running, some lines will show at the bottom. Some of them are empty, and some are just a static copy of one of those progress bars. I am using the newest version, and codes are run on bash.
![snipaste_2019-02-26_21-10-04](https://user-images.githubusercontent.com/22748802/53414958-ece49c80-3a0a-11e9-82bf-8caccd17a2c6.png)

It seems that the update of some progress bar are printed out before the cursor gets back to its original point. I am not familiar with the implementation of tqdm, so I am just guessing.  Hope for your generous help.

**Comment by yelantingfeng** [comment]
Should we always use lock manually at current version? Or lock is already in the implementation of tqdm and this is caused by some other issue?

**Comment by casperdcl** [comment]
I think manual locking may be required (you share a `lock` amongst multiple threads or processes, and call `tqdm.set_lock(lock)` in each one)
